### PR TITLE
feat(auth): add OAuth provider protocol adapters

### PR DESCRIPTION
**Pull request by ColeMurray**
## Summary

- add the static `web` OAuth client authorization service with exact redirect validation, S256-only PKCE, upstream verifier generation, and Google nonce generation
- add GitHub App and Google OIDC provider adapters with runtime-validated identity evidence and bounded public failures
- use exact-pinned `oauth4webapi@3.8.6` for OIDC discovery, token response validation, JWKS signature verification, issuer/audience/expiry checks, and PKCE exchange
- encode provider-indexed request, result, registry, and consumed-flow contracts so Google and GitHub invariants are compile-time enforced
- move OAuth flow and provider credential contracts into the auth layer so protocol code depends on storage ports rather than D1 implementations

## Security properties

- validates client, response type, exact redirect URI, state, PKCE challenge, method, and provider before generating secrets or performing provider/storage work
- enforces the canonical current GitHub and Google issuers
- requests no classic GitHub OAuth scopes; validates GitHub token, user, and every email response at runtime
- normalizes and deduplicates verified GitHub email evidence while preserving a separately selected verified primary email
- exhausts GitHub email pagination with a 10-page bound and fails closed with bounded provider errors on malformed, ambiguous, repeated, cross-origin, or over-limit pagination
- validates Google ID-token structure, signature, issuer, audience, expiry, optional `azp`, and a constant-time comparison against the persisted nonce hash
- keeps provider response bodies and tokens out of errors and logs

## Scope and merge safety

This is an inert protocol/initiation slice of Package 3. It adds no routes, environment bindings, callback handler, identity/admission decisions, session issuance, cookies, or cutover behavior. Existing authentication behavior is unchanged until a later PR composes these modules into the router.

Deferred to the next slices:

- GitHub permission/admission preflight
- provider callback orchestration and immutable identity resolution
- authorization-code issuance/redemption and revocation routes
- rate-limit binding and runtime configuration

## TDD and validation

Implemented with red-green-refactor tracer bullets covering authorization initiation, provider exchange, pagination failure modes, issuer rejection, signature validation, nonce binding, audience, expiry, and authorized-party checks. Review regressions for malformed pagination URLs and email normalization were also reproduced red before their fixes.

- `npm run test -w @open-inspect/control-plane` — 137 files / 2,146 tests
- `npm run test:integration -w @open-inspect/control-plane` — 59 files / 694 tests
- `npm run format:check`
- `npm run lint`
- `npm run typecheck`
- `npm run build`

## Summary by CodeRabbit

* **New Features**
  * Added an end-to-end OAuth authorization flow service for Google and GitHub, generating PKCE-based authorization URLs.
  * Implemented provider-specific handling, including Google OIDC nonce binding and Google identity claim validation.
  * Added typed OAuth flow-state persistence and provider-specific identity/credential mapping (including verified email selection for GitHub).
  * Added defensive request/response validation and a default provider request timeout.
* **Tests**
  * Expanded Vitest suites for authorization validation, Google OIDC exchange/nonce rejection cases, GitHub email pagination edge cases, and typing contracts.

### CODE REVIEW

File: packages/control-plane/src/auth/provider-credential-cipher.ts [hunk 1]
```diff
@@ -3,6 +3,25 @@ export type ProviderCredentialKind =
   | "access_only_expiring"
   | "access_only_nonexpiring";
 
+/** Plain provider credential produced by an authenticated provider adapter. */
+export type ProviderCredentialInput =
```
[the remark this answers is not in the document]
Comment (ColeMurray) [reply packages/control-plane/src/auth/provider-credential-cipher.ts]: Done in e4cfc8b3. ProviderCredentialInput and ProviderCredentialKind now have one neutral auth-domain owner. Provider adapters, storage, and the cipher port import that contract directly, and the DB compatibility re-export has been removed.

File: packages/control-plane/src/auth/providers/github.ts [hunk 2]
```diff
@@ -0,0 +1,318 @@
...
+      .filter((match) => match[2].split(/\s+/).includes("next"));
+    if (nextLinks.length === 0) return null;
+    if (nextLinks.length !== 1) {
+      throw new OAuthProviderError(
+        "malformed_response",
+        "GitHub returned ambiguous email pagination"
+      );
+    }
+
+    const url = new URL(nextLinks[0][1]);
```
[the remark this answers is not in the document]
Comment (ColeMurray) [reply packages/control-plane/src/auth/providers/github.ts]: Fixed in 38b0506e. Malformed pagination URLs are now caught and mapped to OAuthProviderError("malformed_response"). A regression test covers the previous raw TypeError path.

File: packages/control-plane/src/auth/oauth-flow-state.ts [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+      readonly provider: "github";
+      readonly oidcNonceHash: null;
+    })
+  | (ConsumedOAuthFlowStateBinding & {
+      readonly provider: "google";
+      readonly oidcNonceHash: string;
+    });
+
+export interface OAuthFlowStateReader {
+  consume(state: string, expectedProvider: SignInProvider): Promise<ConsumedOAuthFlowState>;
```
[the remark this answers is not in the document]
Comment (ColeMurray) [reply packages/control-plane/src/auth/oauth-flow-state.ts]: Done in e4cfc8b3. consume<P>() now returns the exact consumed-flow variant for expectedProvider. The only assertion remains inside the store after the runtime provider-equality check, so callback callers receive Google's non-null nonce hash without branching or casting.

File: packages/control-plane/src/auth/providers/types.ts [hunk 4]
```diff
@@ -0,0 +1,68 @@
+import type { SignInProvider } from "../sign-in-provider";
+import type { ProviderCredentialInput } from "../provider-credential-cipher";
+
+export interface ProviderAuthorizationRequest {
+  readonly state: string;
+  readonly codeChallenge: string;
+  readonly oidcNonce?: string;
```
[the remark this answers is not in the document]
Comment (ColeMurray) [reply packages/control-plane/src/auth/providers/types.ts]: Done in e4cfc8b3. Provider requests, results, adapters, and the registry are now indexed by SignInProvider. Google nonce bindings are required, GitHub nonce fields are excluded, and compile-time regression tests prevent cross-provider registration or malformed request shapes.

File: packages/control-plane/src/auth/providers/github.ts [hunk 5]
```diff
@@ -0,0 +1,318 @@
...
+    const verifiedEmails = emailEntries.filter((entry) => entry.verified);
+    return {
+      identity: {
+        provider: this.provider,
+        issuer: GITHUB_ISSUER,
+        subject: String(user.id),
+        login: user.login,
+        displayName: user.name ?? user.login,
+        ...(user.avatar_url ? { avatarUrl: user.avatar_url } : {}),
+        verifiedEmails: verifiedEmails.map((entry) => entry.email),
```
[the remark this answers is not in the document]
Comment (ColeMurray) [reply packages/control-plane/src/auth/providers/github.ts]: Fixed in 38b0506e. Verified email evidence is now lowercased and deduplicated before primary-email selection, preserving consistent membership semantics.

**Comment by ColeMurray** [comment]
@open-inspect[bot] Please re-review this PR. The requested changes have been addressed, the review threads are resolved, and the latest CI checks pass.
